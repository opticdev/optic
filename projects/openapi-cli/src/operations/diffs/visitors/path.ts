import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { PathComponents, PathComponent } from '../..';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
