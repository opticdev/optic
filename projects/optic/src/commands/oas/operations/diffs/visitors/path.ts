import { OpenAPIV3 } from '../../../specs';
import {
  OperationDiffResult,
  OperationDiffResultKind,
} from '../../../../capture/patches/patchers/spec/types';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
