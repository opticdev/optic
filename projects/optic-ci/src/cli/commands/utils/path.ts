import path from 'path';
import { promisify } from 'util';
import { exec as callbackExec } from 'child_process';
const exec = promisify(callbackExec);

export const getRelativeRepoPath = (
  relativePath: string,
  gitRootPath: string | false
): string => {
  const absolutePath = path.join(process.cwd(), relativePath);

  if (!gitRootPath) {
    return absolutePath;
  } else {
    return absolutePath.startsWith(gitRootPath)
      ? absolutePath.replace(gitRootPath, '')
      : absolutePath;
  }
};

export const getGitRootPath = async () => {
  const { stdout: gitRootUntrimmed } = await exec(
    'git rev-parse --show-toplevel'
  );
  const gitRoot = gitRootUntrimmed.trim();
  return gitRoot;
};
