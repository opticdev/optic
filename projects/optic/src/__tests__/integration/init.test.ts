import { runOptic, setupWorkspace, fileExists } from './integration';
import path from 'node:path';

jest.setTimeout(30000);

describe('init', () => {
  test('basic init', async () => {
    const workspace = await setupWorkspace('init/basic');

    expect(await fileExists(path.join(workspace, 'optic.yml'))).toBe(false);
    const { combined, code } = await runOptic(workspace, 'init', false);
    expect(await fileExists(path.join(workspace, 'optic.yml'))).toBe(true);
    expect(combined.replace(workspace, '$$workspace$$')).toMatchSnapshot();
    expect(code).toBe(0);
  });

  test('init with existing optic.yml', async () => {
    const workspace = await setupWorkspace('init/existing');

    const { combined, code } = await runOptic(workspace, 'init', false);
    expect(combined.replace(workspace, '$$workspace$$')).toMatchSnapshot();
    expect(code).toBe(1);
  });
});
