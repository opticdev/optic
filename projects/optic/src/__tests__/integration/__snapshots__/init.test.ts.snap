// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init basic init 1`] = `
"Initializing Optic...
Detecting OpenAPI specs in $workspace$...

Found 1 valid spec

Adding files:
- path: example-api-v0.json
  id: Empty

Adding ruleset:
- breaking-changes

File IDs are stable identifiers for your API specifications that will appear in Optic.
You can change them now before you check in the optic.dev.yml file.
"
`;

exports[`init init with \`--ci github-action\` option 1`] = `
"Initializing Optic...
Detecting OpenAPI specs in $workspace$...

Found 1 valid spec

Adding files:
- path: example-api-v0.json
  id: Empty

Adding ruleset:
- breaking-changes

File IDs are stable identifiers for your API specifications that will appear in Optic.
You can change them now before you check in the optic.dev.yml file.

Setting up CI:
- Github action file written to $workspace$/.github/workflows/optic-ci.yml

Set your OPTIC_TOKEN Github secret to finish configuring Optic CI.
More details at: https://github.com/opticdev/github-action/
"
`;

exports[`init init with \`--ci-only github-action\` option 1`] = `
"Deprecation warning: optic.yml file is deprecated. Please rename your file to optic.dev.yml
Initializing Optic...

Setting up CI:
- Github action file written to $workspace$/.github/workflows/optic-ci.yml

Set your OPTIC_TOKEN Github secret to finish configuring Optic CI.
More details at: https://github.com/opticdev/github-action/
"
`;

exports[`init init with existing optic.init.yml 1`] = `
"Deprecation warning: optic.yml file is deprecated. Please rename your file to optic.dev.yml
Error: a configuration file already exists at $workspace$/optic.yml.
"
`;
